### ISSUE TITLE
Prepare for coroutines

### ISSUE THREAD

**Issue by overlookmotel** [sequelize/sequelize#6116]
Now that we're all ES6 for Sequelize v4, the door is open to be using coroutines. Personally, I think this is a great thing - it'll simplify Sequelize's codebase a _lot_.

However, I don't think we're quite ready to do that - I'd like to conclude my research on bluebird and CLS, to make absolutely sure that introducing coroutines won't break CLS support, before we take the plunge.

But when we do start using coroutines, it will be a semver major change, due to a change in the way errors are handled.
#### Error handling

At present some Sequelize async methods throw synchronously if the arguments they are passed are illegal in some way. Once those methods become coroutines, this wouldn't be the case any more - they'll return rejected promises instead.

How it is now:

``` js
Model.findAll = function(options) {
    if (!options) throw new Error('You must provide options!');
}

// Model.findAll() throws synchronously
try {
   Model.findAll();
} catch (err) {
   console.log('It throws!');
}
```

How it will be with coroutines:

``` js
Model.findAll = co.wrap( function*(options) {
    if (!options) throw new Error('You must provide options!');
} );

// Model.findAll() returns a rejected promise
Model.findAll().catch( function(err) {
    console.log('It returns a rejected promise');
} );
```
#### A proposal to "prepare the way"

I would like to see coroutines in Sequelize as soon as possible, but I don't want to hold up the release of v4 until the CLS issues are resolved.

I propose that we "prepare the way" now by converting all Sequelize methods from throwing synchronously to returning rejected promises.

This isn't hard to do using `Promise.method()`:

``` js
Model.findAll = Promise.method( function(options) {
    if (!options) throw new Error('You must provide options!');
} );

// Model.findAll() returns a rejected promise
Model.findAll().catch( function(err) {
    console.log('It returns a rejected promise');
} );
```

This way:
1. The whole API switches away from sync throwing in one go.
2. It can all be done as a single change in Sequelize v4.0.0.
3. Methods can then be refactored with coroutines as and when we feel like it, without creating any further breaking changes.
4. We won't have to wait until Sequelize v5 to start using coroutines.

What do you think?

**Comment by mickhansen** [comment]
We're in 4.0.0-pre so it's the time for breaking changes!
4.0.0 is about cleaning up all the code we want so we can be more productive.

**Comment by overlookmotel** [comment]
@mickhansen So that's a yes? :)

**Comment by overlookmotel** [comment]
@felixfbecker Are there any major ES6 changes still in the works which might interfere with this if I start putting together a PR?

**Comment by mickhansen** [comment]
That's a we're introducing co routines before 4.0.0 launches.
See other issues, we'll likely use function *() { in classes and then just a post processor to rebind and wrap.

**Comment by overlookmotel** [comment]
My opinion is that we're not ready for coroutines quite yet.

Please see my comment here https://github.com/sequelize/sequelize/issues/5898#issuecomment-221346901 about coroutines being responsible for loss of CLS context in some cases.

My feeling is that ensuring CLS support is 100% robust is really important because it's used for running queries in the right transactions - so a bug threatens to undermine consistency, which I think is a pretty big deal.

I'm going for it fast as I can in trying to iron out the CLS/bluebird issues. But it's really complex (see [test suite I'm building](https://github.com/overlookmotel/cls-bluebird-test) and [issue on CLS](https://github.com/othiym23/node-continuation-local-storage/issues/64)), so it may take a while and might not be ready for Sequelize v4 if that's imminent.

Of course, I don't have the final say in these things! But in my opinion, keeping CLS support bug-free is critical.

**Comment by overlookmotel** [comment]
By the way, @mickhansen if the post-processor you mention is https://github.com/danielstjules/async-class, that looks like a nice clean way to co-wrap the methods of a class.

**Comment by felixfbecker** [comment]
@overlookmotel No, not really. Only the tests and template strings.

Personally I'm -1 on coroutines because
 - they can't be defined as methods currently without calling a post processor, which sucks for static analysis, V8 optimization (it's not an optimized static class anymore because you modify the whole class afterwards, so V8 creates hidden classes), etc...
 - generators cannot be optimized at all currently by V8. If you are using generators, the compiler instantly switches to non-optimized mode and there's nothing you can do.
 - I don't think the code is that bad now that we can use arrow functions in promises.

**BUT** I'm totally 100% for always either throwing sync or async, but never both. It releases Zalgo. So go ahead with this "always throw async" thing, and maybe we can do the important stuff like cleaner API for ES6 users, connection-independent models etc. first and later use coroutines after 4.0 is released.

**Comment by mickhansen** [comment]
@felixfbecker We will need to evaluate the performance of coroutines via our benchmark suite in any case, but having used `async/await` a lot in professional projects it's something me and @janmeier think would increase contributor productivity greatly.

**Comment by janmeier** [comment]
Definitely agree that we should wrap everything in promise.method to be consistent with regards to sync / async errors.

It does however have the same problems as wrapping in a coroutine does - The class is no longer static so V8 might de-opt it. Of course we could wrap the body of each method instead, but pretty it is not.

**Comment by felixfbecker** [comment]
@janmeier `Promise.method()` is not needed, simply replace all occurences of `throw new Whatever()` with `return Promise.reject(new Whatever())` (of course only outside of promise handlers). Or start every method with `return Promise.try()`, as it is already used in the codebase in quite a few places.

**Comment by janmeier** [comment]
Regarding CLS, we might also consider if the extra :monkey: patching work is worth it, compared to a switch to domains. Yes I know domains are deprecated, but at least they are officially supported by bluebird.

I'm not saying its a good solution, but I don't think there are any good solutions

**Comment by felixfbecker** [comment]
We could also pass the contexts _explicitely_ via options. I always prefer explicit over implicit and really dislike CLS in general because it is so magic.

**Comment by overlookmotel** [comment]
@felixfbecker Contexts are passed explicitly with options *inside* Sequelize. The CLS support is for 2 things: (1) so that if the user wants to use CLS in their own code, calling a Sequelize method doesn't foul it up and (2) transaction can be passed via CLS.

@janmeier Personally I'm quite invested in CLS so I'm going to do the work of working out the monkey-patching anyway. It's just a question of how long it'll take.

The replacement for both domains and CLS is under development in node core. [Async Wrap](https://github.com/nodejs/diagnostics/blob/master/tracing/AsyncWrap/README.md) is actually already in node, but undocumented, and it probably won't be too long until it's officially released.

So I don't think now is a particularly good time to switch to domains.

**Comment by felixfbecker** [comment]
@overlookmotel If we need to patch global Bluebird, the recently added [`Promise.getNewLibraryCopy()`](http://bluebirdjs.com/docs/api/promise.getnewlibrarycopy.html) might be helpful

**Comment by felixfbecker** [comment]
Lol just saw the changelog :D

**Comment by janmeier** [comment]
Didn't know about async wrap - In that case its definitely not worth it to try to replace CLS right now

**Comment by Verdier** [comment]
Just a comment to expose how we use co in our project with class. We have build a small tool that take the prototype and wrap all generator methods:

```js
class MyClass {
 mySync() {}
 * myAsync() {}
}

module.exports = wrapProto(MyClass);
```
Here is the implementation:
```js
'use strict';

const co = require('co');

module.exports = function (Ctr, wrap) {
    wrap = wrap || co.wrap;

    let proto = Ctr.prototype;
    wrapOwnProperties(proto, wrap);
    wrapOwnProperties(Ctr, wrap);

    return Ctr;
};

function wrapOwnProperties(obj, wrap) {
    let names = Object.getOwnPropertyNames(obj);

    for (let name of names) {
        let descriptor = Object.getOwnPropertyDescriptor(obj, name);
        if (descriptor.get || descriptor.set) continue;

        if (obj[name].constructor && obj[name].constructor.name === 'GeneratorFunction') {
            obj[name] = wrap(obj[name]);
        }
    }
}
```
An other point concerning `Promise.method` and `throw` vs `return Promise.reject()`: it's better to user `Promise.method` to be ready for co. Because using co, you will have to throw, and you will need to rewrite all your `Promise.reject()`...

**Comment by felixfbecker** [comment]
@Verdier but `Promise.method()` is not possible with class methods, you have to assign on the prototype. Changing that back to classes for co is a much bigger diff than search-and-replacing `return Promise.reject(.+)` with `throw $1`

**Comment by Verdier** [comment]
What about using our trick for wrapping with co, but to wrap with `Promise.method()`?

```js
'use strict';

const co = require('co');

module.exports = function (Ctr) {
    wrapOwnProperties(Ctr.prototype);
    wrapOwnProperties(Ctr);
    return Ctr;
};

function wrapOwnProperties(obj) {
    let names = Object.getOwnPropertyNames(obj);

    for (let name of names) {
        let descriptor = Object.getOwnPropertyDescriptor(obj, name);
        if (descriptor.get || descriptor.set) continue;

        if (obj[name].constructor && obj[name].constructor.name === 'GeneratorFunction') {
            obj[name] = Promise.method(obj[name]); // <---- HERE 
        }
    }
}
```

**Comment by overlookmotel** [comment]
@Verdier @felixfbecker I guess `Promise.try()` at the very start of every function body would cover both needs. Inside the `Promise.try()` body, you `throw`. But without wrapping in `Promise.method()`.

That said, if everything's going to have to be wrapped in `co.wrap()` (or similar) for co-routines, it'd be good to check whether whether wrapping does indeed cause a performance impact.

**Comment by overlookmotel** [comment]
@Verdier [async-class](https://www.npmjs.com/package/async-class) appears to do the same as your wrapping implementation. I've got a PR in on it at the moment to allow you to specify your own wrapper.

**Comment by overlookmotel** [comment]
Hi all. A bit of an update:

### cls-bluebird

I'm making pretty good progress with fixing cls-bluebird. I've got what I think is a working product now, and am most of the way through creating a huge test suite that absolutely drills it from every possible angle to make sure it's really robust (34,000 tests so far!).

Here's my work in progress: https://github.com/overlookmotel/cls-bluebird/tree/refactor

If anyone wants to do a code review, that'd be very welcome.

I estimate it'll be complete in a couple of weeks.

### async-class

For wrapping all class methods which are generators, [async-class](https://www.npmjs.com/package/async-class) looks like a good solution to me.

I've got a PR in progress to allow you to specify a wrapper, which will make it possible to use `Sequelize.Promise.coroutine`: https://github.com/danielstjules/async-class/issues/4

@felixfbecker @Verdier can one of you explain why there's a problem with wrapping class methods with `Promise.method()`? I'm not au fait with ES6 classes, so I don't get it...

**Comment by mickhansen** [comment]
@overlookmotel the machine

**Comment by overlookmotel** [comment]
@mickhansen ???

In truth I have doubted my sanity in the lengths I've ended up going to get cls-bluebird working (and the stupid amount of tests). But it's one of those things where once you've got half way, you might as well continue to the end...

**Comment by mickhansen** [comment]
@overlookmotel the @ was meant as a part of the sentence as in "you're a machine" :)

**Comment by overlookmotel** [comment]
@mickhansen I figured that but I wasn't sure if it was a compliment! (I'm still not sure and I still doubt my sanity)

**Comment by mickhansen** [comment]
@overlookmotel It was, sort of :)

**Comment by overlookmotel** [comment]
After some further discussion about coroutines on #6266, here are my overall thoughts on adoption of coroutines:

### Benefits of coroutines

I am very pro coroutines.

NB what I'm talking about is using `Bluebird.coroutine()` - no need to add [co](https://www.npmjs.com/package/co) as a dependency.

I feel that using coroutines will make it easier + faster to write code within Sequelize and make the codebase much more readable.

I don't feel that this would be offputting to contributors - coroutines are pretty mainstream now. If anything they're easier for the uninitiated to understand as opposed to all the lesser-known methods of Bluebird (`.tap()`, `.return()` etc) which are at present used throughout Sequelize and are hard to reason about unless you know Bluebird *really* well.

And now that [cls-bluebird](https://www.npmjs.com/package/cls-bluebird) new version has been released, CLS support is no longer a problem.

### Wait until `async`/`await`?

@felixfbecker has said that `async`/`await` will be supported in node v8 which is due release in April 2017 (6 months time). So one approach could be not to use coroutines at all and just wait for `async`/`await`.

I don't think we should do that. Using `async`/`await` would require dropping support for all node versions < v8.x. This which would be a semver major change and also probably undesirable until pretty much all users are on v8.x anyway (NB Node v6.x won't leave LTS until April 2019!).

So we could be waiting a couple of years before we can use `async`/`await`. In my opinion, that is just too long to wait.

Also, at the moment, native promises are much less performant than Bluebird, so using `async`/`await` could lead to a big performance hit.

### Babel transpilation

This has already been ruled out by @mickhansen. And personally I agree.

### Adoption over time

There's no need to refactor the whole codebase now to coroutines - it can just be done as and when people want to, probably when they're refactoring a function anyway.

### Performance

On the performance question: what has been established so far is that Bluebird coroutines perform roughly the same as Bluebird promises (in some cases they're faster). What we *don't* know is whether there's a performance hit from defining class static methods and prototype methods as generators and then using something like [async_class](https://www.npmjs.com/package/async-class) to convert them to coroutines.

This would look like this:

```js
class Model {
    *findAll() {
        yield ...
    }
}

require('async-class').wrap( Model );
```

I guess only way to determine if this causes a V8 deoptimization is to do some tests. I'm not really a performance guy and I don't know benchmarking tools at all. So I wonder if someone else more familiar with this area would like to do a benchmark?

Or @felixfbecker could you explain in a bit more detail what your concerns are here?

### What needs to be done now

Regardless of the outcome of this discussion, what I do feel strongly is we should leave the door open to adopting coroutines at some point during the life of Sequelize v4.x without breaking changes.

This does require one breaking change - converting all the API to be promise-returning in all cases. i.e. no synchronous throwing and return rejected promises instead. This will allow adoption of coroutines when/if ever we feel like it without any breaking changes.

**Comment by felixfbecker** [comment]
V8 optimization. Generators are not optimzied at all by V8 currently: https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#2-unsupported-syntax

Also, V8 creates hidden classes for objects. Every time you assign a property to a property, it throws the old hidden class away and creates a new one. The benefit of native classes is that the class is declared upfront and we don't have to assign on the prototype anymore. When you wrap with async-class, it will patch every single method with a new wrapper method.

**Comment by mickhansen** [comment]
> When you wrap with async-class, it will patch every single method with a new wrapper method.

But presumably this would only happen once per class per startup - Would it have any perf impacts in long running code?

**Comment by overlookmotel** [comment]
> Every time you assign a property to a property

Do you mean "assign a property to an *object*"?

As far as I'm aware from what I've read (sorry can't remember the reference), coroutine code ran roughly as fast as promises. Sometimes faster, sometimes slower dependent on amount of parallelism. This result was also what a friend of mine found doing a benchmark of Bluebird promises vs Bluebird coroutines. So it sounds like the lack of V8 optimization isn't a problem. And once V8 does optimize generators, which it says [here](https://groups.google.com/forum/#!topic/v8-users/KnnUb-u4rA8) was meant to happen some time this year, coroutines will probably be faster overall.

What penalty wrapping class methods produces is a separate question, and not one I have any idea of the answer, though I suspected the same as @mickhansen.

What'd be best way of going about finding out for sure?

**Comment by overlookmotel** [comment]
Anyway, I guess my main point is that we should make the changes now in Sequelize v4 that would allow coroutines to be adopted in future without breaking changes. That consists of converting any methods which currently throw synchronously to instead return rejected promises.

Do you guys agree with that?

**Comment by felixfbecker** [comment]
All methods should either always throw sync or always return rejected promises. I don't know if I understood you right, but sync methods should still throw sync.

**Comment by overlookmotel** [comment]
@felixfbecker Sorry yes, I wasn't clear. I'm proposing that:

- Sync methods should always return a value synchronously or throw synchronously
- Async methods should always return a promise, which either resolves with a value or rejects with an error

At present, a few async methods return a promise mostly, but in some circumstances throw synchronously instead. I'm proposing that they be altered to return a rejected promise rather than throwing.

I seem to remember there's one method which sometimes returns a result sync and sometimes returns a promise (but I can't remember which method!). This is a bit of an anti-pattern to my mind, and should probably be split into two methods, one always sync and one always async.

**Comment by felixfbecker** [comment]
Definitely. This releases Zalgo.

**Comment by overlookmotel** [comment]
@mickhansen Do you agree?

**Comment by mickhansen** [comment]
Definitely agree, zalgo via superimposed synchronicity states is bad!

**Comment by felixfbecker** [comment]
Btw, Node 7.4 has native async/await

**Comment by overlookmotel** [comment]
At present that would screw up CLS support - native promises don't support CLS. But once async_hooks stabilises, CLS will use that underlying mechanism instead and that particular problem will go away. Give it a couple of months I'd say.

**Comment by janmeier** [comment]
Really looking forward to async_hooks - Alone for the fact that we will then have something that's stable, and officially supported by core (as opposed to CLS and domains)

**Comment by felixfbecker** [comment]
We will have to wait until Node 4 is out of LTS anyway.
Personally I would never use CLS (or domains), the perf impact is just way to high.

**Comment by overlookmotel** [comment]
@janmeier Yes me too!
@felixfbecker I imagine CLS will be adapted so it uses the old methods on old versions of Node and async_hooks on recent node versions. The performance of async_hooks is meant to be very good, but I haven't benchmarked it yet.

**Comment by overlookmotel** [comment]
I think this issue should remain open.

**Comment by papb** [comment]
@overlookmotel What is the status/ideas on this? After two years, anything has changed? New ideas?

**Comment by sushantdhiman** [comment]
Superseded by https://github.com/sequelize/sequelize/issues/11859, Closing this issue
